contracts/libraries/Welford.sol

```diff
@@ -10,60 +10,66 @@ library Welford {
+     * @notice Performs an update of the mean and stdev using online algorithm
      * @param curCount is the current value for count
-     * @param curMean is the current value for mean
-     * @param curM2 is the current value for M2
+     * @param oldValue is the old value to be removed from the dataset
      * @param newValue is the new value to be added into the dataset
+     * @param curMean is the current value for mean
+     * @param curDSQ is the current value for DSQ
      */
     function update(
         uint256 curCount,
+        int256 oldValue,
+        int256 newValue,
         int256 curMean,
-        uint256 curM2,
-        int256 newValue
-    )
-        internal
-        pure
-        returns (
-            uint256 count,
-            int256 mean,
-            uint256 m2
-        )
-    {
-        int256 _count = int256(curCount + 1);
-        int256 delta = newValue.sub(int256(curMean));
-        int256 _mean = int256(curMean).add(delta.div(_count));
-        int256 delta2 = newValue.sub(_mean);
-        int256 _m2 = int256(curM2).add(delta.mul(delta2));
-
-        require(_count > 0, "count<=0");
-        require(_m2 >= 0, "m2<0");
+        int256 curDSQ
+    ) internal pure returns (int256 mean, int256 dsq) {
+        // Source
+        //https://nestedsoftware.com/2019/09/26/incremental-average-and
+        //-standard-deviation-with-sliding-window-470k.176143.html
+        if (curCount == 1 && oldValue == 0) {
+            // initialize when the first value is added
+            mean = newValue;
+        } else if (oldValue == 0) {
+            // if the buffer is not full yet, use standard Welford method
+            int256 meanIncrement =
+                (newValue.sub(curMean)).div(int256(curCount));
+            mean = curMean.add(meanIncrement);
+            dsq = curDSQ.add((newValue.sub(mean)).mul(newValue.sub(curMean)));
+        } else {
+            // once the buffer is full, adjust Welford Method for window size
+            int256 meanIncrement = newValue.sub(oldValue).div(int256(curCount));
+            mean = curMean.add(meanIncrement);
+            dsq = curDSQ.add(
+                (newValue.sub(oldValue)).mul(
+                    newValue.add(oldValue).sub(mean).sub(curMean)
+                )
+            );
+        }
 
-        count = uint256(_count);
-        mean = _mean;
-        m2 = uint256(_m2);
+        require(dsq >= 0, "dsq<0");
     }
 
     /**
      * @notice Calculate the variance using the existing tuple (count, mean, m2)
      * @param count is the length of the dataset
-     * @param m2 is the sum of square errors
+     * @param dsq is the variance * count
      */
-    function variance(uint256 count, uint256 m2)
+    function variance(uint256 count, int256 dsq)
         internal
         pure
         returns (uint256)
     {
         require(count > 0, "!count");
-        return m2 / count;
+        return uint256(uint256(dsq) / count);
```

Comment:
Doing `uint256(dsq)` will actually give some [undefined results](https://github.com/ethereum/solidity/issues/534) if the int is negative. We need to check and revert

```solidity
require(dsq > 0, "dsq is negative or zero");
```

Reply:
changed to require (dsq >= 0)